3단계 : 자동차 경주

aeomhs opened the pull request:
안녕하세요. 🙇‍♂️

3단계 자동차 경주 과제 리뷰 요청드립니다. 🙏

`README.md` 파일에 요구사항/구현사항을 미리 적어놓고 진행해보았습니다.   
요구사항에 써있는 `주어진 횟수 동안 n대의 자동차는 전진 또는 멈출 수 있다.` 와 같은 문장을 바로 써두기엔, 체크하기가 쉽지 않을 것 같아서 이런 저런 생각을 해보면서 쪼개어 리스트를 만들어 보았습니다.  

가능한 실패 테스트를 먼저 작성하고 구현을 해보려고했는데, 
처음부터 구체적인 실패 테스트를 작성하기엔 구현에 대한 계획이 없으니 쉽지 않았고, 
또 구현을 해 나아가다보면 테스트 코드를 놓치게 되는 것 같아서
균형을 잡으면서 해보려고 노력했습니다.
재미있는 시간이었습니다 😆

자동차의 움직임 여부를 Random 하게 결정해야한다는 것이 테스트 코드에 약간 어려움을 주었던 것 같습니다.  

처음에는 `Turn` 이라는 개념을 넣어서 한 턴, 한 턴마다 자동차의 움직임을 결정하고 각 턴의 결과를 확인할 수 있도록 해보려고 했는데요,
미션을 보자마자 이후에 코루틴을 사용해서 Racing 이 시작되면 여러 Car를 동시에 racing 하도록 해보고 싶었는데, 이를 위해서는 턴제 방식으로 그려나가는게 조금 어려울 것 같더라구요.
그래서 한 번에 경주를 시작/종료시키고 경주로 인한 Track 에 남은 자동차의 흔적 Trace 를 통해 결과를 그리도록 해보았습니다.

CommandLineInterface .. 즉 뷰쪽은 더 좋은 방법이 있을 것 같은데,, 우선 너무 늦어지는 것 같아서 리뷰 요청을 먼저 드립니다!

미리 소중한 피드백 감사합니다. 🙇‍♂️


### 이전 리뷰목록
- 1단계 : 코틀린 기초 #390 
- 2단계 : 문자열 계산기 #456

aeomhs committed:
refactor: (step-2)regex string literal (triple quote) 방식으로 수정

aeomhs committed:
refactor: (step-2) 연산자별 로직을 Operator enum 내부로 이동

aeomhs committed:
docs: 간단하게 step3 구현사항 정리

aeomhs committed:
test: Track 실패 테스트 작성

Track 은 자동차가 달릴 수 있는 공간이자,
자동차가 달린 흔적이 남아있는 객체입니다.
이후 자동차가 달린 흔적을 추적할 수 있습니다.
- Track::markTrace 를 통해 자동차가 흔적을 남길 수 있습니다.
- Track::isPassedAt 을 통해 자동차의 흔적을 추적할 수 있습니다.

aeomhs committed:
test: Car 실패 테스트 작성

Car 는 경주용 자동차입니다.
오직 Track 위에서 달릴 수 있습니다.
- Car::raceOn 을 통해 트랙에서 moveCount 만큼 움직일 수 있습니다.

aeomhs committed:
test: Racing 실패 테스트 작성

Racing 는 자동차 경주 한 경기에 대한 객체입니다.
외부 인터페이스에 의해 생성하기 위한 new 함수와
경주를 시작하기 위한 start 메서드를 제공합니다.

aeomhs committed:
docs: 자동차가 임의로 움직이기 위한 엔진 개념 추가

자동차가 임의 확률로 전진하는 것에 대한 테스트를 어떻게 하는 것이 좋을지 고민하다보니, Random 객체를 의존주입해주어야 할 필요성을 느꼈고,
Random 객체를 포함하는 엔진이라는 개념을 자동차에 추가.

aeomhs committed:
test: Random 객체 학습 테스트 코드 추가

Random 객체를 mock 주입하기위한 학습 테스트 추가

aeomhs committed:
style: klintFormat 적용 .

음.. 이상하게 gitHook 이 동작하지 않는데.. ㅠㅠ

aeomhs committed:
test: mockk 라이브러리 추가

aeomhs committed:
test: Car 실패 테스트 코드 수정

실제 요구사항에 맞게 Random 하게 움직여야 한다는 내용을 추가

aeomhs committed:
feat: Track 구현

aeomhs committed:
test: Track 예외 처리 테스트 추가

aeomhs committed:
feat: Car 구현 및 테스트 오류 수정

Random mocking 을 한 뒤에 Car 객체 생성해야 테스트 가능.

aeomhs committed:
docs: 자동차, 트랙 구현

aeomhs committed:
test: Racing 결과 조회를 위한 테스트 추가

aeomhs committed:
test: Racing 테스트 케이스 수정

여러 개의 테스트 케이스로 나누기 애매하여.. 합침!

aeomhs committed:
test: Racing 테스트 케이스 수정 및 Car id 추가

Racing 구현을 위한 테스트 케이스 수정

aeomhs committed:
feat: Racing 구현

aeomhs committed:
bug: 실행마다 같은 runtime & seed 로 인해 같은 랜덤 결과나오는 문제 해결

랜덤...에 점점... 빠져드는 느낌

aeomhs committed:
feat: 간단한 CommandLineInterface 구현

aeomhs committed:
docs: 구현 사항 README 적용

aeomhs committed:
refactor: Track::isPassedAt 제거 Track::traces 구현

isPassedAt 메서드를 사용함으로써 Track 내부에 Array(traces)에 의존하게되고,
순서에 대한 의존이 발생했다.
Track 을 사용하는 곳에서 0번째부터인지 1번째부터인지 알아야하는 문제가 발생하였다.
Iterator 로 바꾸어 리턴하도록 수정하여 Track 내부의 Array<Trace> 에 의존하지 않도록 수정했다.

aeomhs committed:
refactor: 패키지 interfaces / racing 분리

aeomhs committed:
refactor: 일정한 순서를 위해 LinkedHashMap 사용

aeomhs committed:
refactor: IntRange -> for loop

aeomhs committed:
docs: README.md 구현 내용 체크

aeomhs committed:
refactor: IntRange -> for loop

karian7 commented on src/main/kotlin/camp/nextstep/edu/step2/StringCalculator.kt:
```diff
@@ -12,8 +12,8 @@ package camp.nextstep.edu.step2
-    private val leftOneRegex: Regex = Regex("(-?\\d+)")
-    private val rightsRegex: Regex = Regex("([-+/*])\\s?(-?\\d+)")
+    private val leftOneRegex: Regex = Regex("""(-?\d+)""")
+    private val rightsRegex: Regex = Regex("""([-+/*])\s?(-?\d+)""")
```
👍

karian7 commented on src/main/kotlin/camp/nextstep/edu/step2/StringCalculator.kt:
```diff
@@ -39,11 +39,6 @@ object StringCalculator {
-            Operator.TIMES -> (left.times(right))
-            Operator.DIVIDE -> (left.div(right))
-        }
+        return operator.operate(left, right)
```
Nit: 연산자 오버로딩을 이용하면 바로 함수로서 사용도 가능해집니다.
```suggestion
        return operate(left, right) // Operator 객체의 변수명을 operate 로 받았음
```

참고: https://wooooooak.github.io/kotlin/2019/03/21/kotlin_invoke/

karian7 commented on src/main/kotlin/camp/nextstep/edu/step3/README.md:
```diff
@@ -0,0 +1,77 @@
+    - [ ] 한 번의 경주에는 N대의 자동차가 경주에 참여할 수 있다.
+- [ ] 사용자 인터페이스 
+  - [ ] 사용자는 레이싱 경주에 필요한 N, M 정수를 입력할 수 있다.
+  - [ ] 레이싱 결과를 확인할 수 있다.
```
요구사항 정리 👍 
처음엔 쉽지 않아요. 실제로 만들다 보면 여러 테스크가 한번에 되거나 새로운 테스크가 추가되기도 합니다.
익숙해 익숙해 지면 예외케이스등 암묵적인 요구사항들을 발견할수도 있고 더 나가면 추정 정확도를 높이는데도 도움이 됩니다.
게임하듯이 한개씩 클리어 해 나가는 재미가 생기는건 덤이고요.ㅎㅎ 🎲

karian7 commented on build.gradle.kts:
```diff
@@ -13,6 +13,7 @@ repositories {
 dependencies {
     testImplementation("org.junit.jupiter", "junit-jupiter", "5.8.2")
     testImplementation("org.assertj", "assertj-core", "3.22.0")
+    testImplementation("io.mockk", "mockk", "1.12.3")
```
Random 을 static mocking 하기 위해 추가하신것 같아요. 그런데 이 교육과정에서는 적절하게 경계를 나누면 mockito 가 필요 없어요. mockito 의존 없이 도전해 보시겠어요?

karian7 commented on src/main/kotlin/camp/nextstep/edu/step3/Track.kt:
```diff
@@ -1,12 +1,20 @@
 
     fun markTrace() {
-        TODO()
+        if (onTraceIndex >= length) throw RuntimeException("end of track; cannot mark trace more.")
```
[check](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/check.html) 함수의 용법을 확인하고 활용해 볼까요?

karian7 commented on src/main/kotlin/camp/nextstep/edu/step3/Track.kt:
```diff
@@ -1,12 +1,20 @@
 package camp.nextstep.edu.step3
 
-class Track(length: Int) {
+class Track(private val length: Int) {
```
`traces: Array<Trace>` 를 감싸는 객체를 만드셨네요. 좋은 접근 입니다. 
`Car` 객체 관점에서 `Array<Trace>` 를 직접 의존하지 않고 `Track` 객체를 의존하도록 했을 때 어떤 장점이 있을까요.

aeomhs replied:
사실,, 처음부터 Trace 라는 개념을 생각했다기보다,
자동차가 달릴 수 있는 Track 을 먼저 생각했었습니다.
그러다가 어떻게 자동차가 달린 결과를 상태로 관리할 수 있을까 고민해보다가
자동차가 달리고 난 뒤의 흔적을 관리하면 될 것 같다는 생각에 Trace 라는 개념을 생각했고,
Trace 는 Track 이 외에는 굳이 알 필요가 없다고 생각했었습니다. (마치 ,, LinkedList 의 Node 처럼요)

Trace 의 상태 변화는 꼭 필요한 상태였고, (marked 는 가변이므로) 
Trace 의 상태 변화는 Track 에서만 관여하도록 강제하고 싶었습니다. 

생각해보진 않았지만,, Car 객체가 Array<Trace> 에 직접 의존하게된다면,, (Track 없이)
Trace 상태를 가져오기 위해서는 Car 를 통해 가져와야겠네요
Trace 의 형태가 다양해지면 Array 대신 다른 자료형을 사용해야 할 수도 있을 것 같습니다. 
Car 객체는 점점 비대해지고,, 복잡해질 것 같습니다.

음... 잠깐 생각해보았지만, 수치나, 명확한 근거가 있는 장점이 생각나진 않습니다..
그런데 자동차가 자신의 흔적을 관리하는 것보단,,
자동차가 자신이 달린 트랙에 흔적을 남기는 것이 조금 더 자연스러운 것 같습니다 🤔

뭔가.. 한참 고민해보았네요 ㅎㅎㅎ
좋은 질문주셔서 감사합니다 😆

karian7 replied:
우와.. 도메인에 대한 관찰력과 추상화 능력이 뛰어나시군요!! 👏👏 
짧은 질문에 깊은 생각을 해 주신것 같습니다.

> 그런데 자동차가 자신의 흔적을 관리하는 것보단,,
> 자동차가 자신이 달린 트랙에 흔적을 남기는 것이 조금 더 자연스러운 것 같습니다

저는 이 말씀이 핵심이라고 생각합니다. `Array` 는 어디서나 사용되는 단순한 도구에 가까운 자료구조이지만 `Track` 은 도메인에 의미있는 특별한 객체 입니다. `Car` 관점에서 `Array` 는 상태를 저장하는 종속적인 관계에 가깝지만 `Track` 은 `Car` 와 동등한 위치에서 협력하는 관계라고 생각합니다. 그리고 도메인 모델을 잘 표현할수록 가독성과 유지보수성이 올라간다고 생각해요. 이 내용은 다음 미션인 로또 만들기에서 좀 더 자세하게 다루고 또한번 훈련하게 됩니다. 😀

karian7 commented on src/test/kotlin/camp/nextstep/edu/step3/CarTest.kt:
```diff
@@ -1,17 +1,43 @@
+        val id = 1
+        val car = Car(id)
+        val mockkRandom = mockk<Random>()
+        mockkStatic("kotlin.random.RandomKt")
```
mock static method는 테스트가 fork 모드에서 실행되지 않는다면 다른 테스트에서 실행하는 로직에서 의존하는 랜덤도 이 테스트에 설정된 값의 영향을 받을 수 있어서 격리싴야 해요. `try with resources` 또는 [use](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html) 함수를 사용해서 격리시킬 수 있어요.

https://javadoc.io/static/org.mockito/mockito-core/4.0.0/org/mockito/Mockito.html#static_mocks

aeomhs replied:
오!!! 좋은 정보 감사합니다!! 👍👍👍

karian7 commented on src/test/kotlin/camp/nextstep/edu/step3/CarTest.kt:
```diff
@@ -1,17 +1,43 @@
+        assertFalse(track.isPassedAt(1))
+        assertFalse(track.isPassedAt(2))
+        assertFalse(track.isPassedAt(3))
+        assertFalse(track.isPassedAt(4))
```
`ObjectEnumerableAssert#allSatisfy(Consumer)` 를 찾아보고 활용해 봅시다.

karian7 commented on src/test/kotlin/camp/nextstep/edu/step3/CarTest.kt:
```diff
@@ -1,17 +1,43 @@
+        val car = Car(id)
+        val mockkRandom = mockk<Random>()
+        mockkStatic("kotlin.random.RandomKt")
+        every { mockkRandom.nextInt(0, 10) } returnsMany listOf(8, 9, 1, 2, 3)
```
테스트 코드도 프러덕션 코드와 마찬가지로 유지보수 대상이라고 생각합니다. 따라서 가독성,간결성이 좋은 테스트 코드를 만들어야 해요. `8`, `9` 같은 숫자를 "이동함" 이라는 의미를 담은 변수 또는 상수로 선언한다면 좀 더 테스트 코드를 이해하는데 도움이 될 것 같습니다. `canRaceOnTrack` 메소드가 긴데 전체적으로 가독성을 고려해서 리펙터링 해 봅시다.

karian7 commented on src/main/kotlin/camp/nextstep/edu/step3/Car.kt:
```diff
@@ -7,12 +7,13 @@ class Car(id: Int) {
     fun raceOn(track: Track, moveCount: Int) {
-        engine.move()
-        TODO()
+        IntRange(0, moveCount).forEach { _ ->
```
사용하지 않는 변수에 대해 `_` 사용 좋네요 👍 
`repeat` 함수도 찾아봅시다.

karian7 commented on src/main/kotlin/camp/nextstep/edu/step3/Car.kt:
```diff
@@ -1,8 +1,18 @@
     }
+
+    private class Engine(id: Int) {
+        private val random = Random(id)
```
mockStatic 사용과 관련해서 이 Random 객체를 제어하는 부분이 관건일 것 같습니다. 엔진에 대한 의존을 생성자 주입으로 바꾸고 프러덕션에서 사용되는 엔진과 테스트 코드에서 사용하는 엔진을 구분해 보면 어떨까요?  "kotlin default argument" 에 대해서도 참고해 보시면 좋겠습니다.

karian7 reviewed (changes requested):
3단계 구현을 재미있게 잘 하신것 같습니다. 😀 이전단계 피드백 반영도 잘 해 주셨습니다. 💯 

`Car` 객체와 `Track` 객체로 나누어 각각 이동과 기록 책임을 주고 협력하게 만든부분도 좋았습니다. 👍 👍

커밋을 잘 나누어서 커밋을 따라가면서 어떤 흐름으로 구현해 나가셨는지 이해할 수 있었습니다. 쉽지 않지만 과정 마칠때까지 연습한다면 실무에서도 커밋을 잘 나눌 수 있는 실력이 되실 수 있다고 생각합니다.

몇가지 피드백 드렸는데요. 커밋을 따라가면서 남긴거라 최종코드에 없어진 부분도 있습니다. 유사한 다른 코드에 반영해 보시거나 없다면 그냥 참고만 해 주세요. 😅 

Random 을 제어하기 위해 mockito 를 사용하신 부분은 이번 단계에서 수정하고 진행하면 좋을것 같아서 Request change 합니다. 그럼 확인해 보시고 다시 요청 바랍니다~

aeomhs committed:
refactor: (step-2) Operator overload invoke !!!

aeomhs committed:
refactor: remove mockk dependency

mockito 의존 제거
Car 외부에서 Engine을 주입하도록 수정.

aeomhs committed:
refactor: rename CarTest tc name

테스트 케이스 네임 수정

aeomhs committed:
refactor: 코틀린 check 함수 사용

aeomhs committed:
refactor: Engine 인터페이스 및 RandomlyEngine 구현체 Car 내부로 이동

밖에 있을 필요가 없어 보임
Engine 의 speed 리턴에 대해 Car 행위가 긴밀하게 의존하고 있는데, 같은 곳에서 보는 것이 좋지 않을까?

aeomhs committed:
refactor: 테스트를 위한 InstantEngine.kt 구현

랜덤하지 않고, 정해진 회수만큼 움직일 수 있는 값을 리턴하는 테스트용 객체

aeomhs committed:
test: Car, Track 에 대한 테스트 코드 추가

aeomhs committed:
test: 테스트용 InstantEngine 테스트 코드 추가

aeomhs reviewed (commented):
기본 Engine 을 요구사항에 맞는 RandomlyEngine 을 사용하도록 하고,
테스트를 위한 InstantEngine 을 추가했습니다.
말씀하신 의도가 잘 반영되었는지 궁금합니다!

karian7 commented on src/main/kotlin/camp/nextstep/edu/step3/racing/RandomlyEngine.kt:
```diff
@@ -0,0 +1,20 @@
+
+import kotlin.random.Random
+
+class RandomlyEngine : Engine {
```
전략 패턴 활용 💯

karian7 commented on src/main/kotlin/camp/nextstep/edu/step3/racing/RandomlyEngine.kt:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    override fun speed(): Int {
+        return if (random.nextInt(0, 10) >= 4) 1 else 0
```
`4`, `0`, `1` 같은 매직넘버들도 적절한 상수화를 통해 의미부여와 가독성을 높여보면 어떨까요.

karian7 commented on src/test/kotlin/camp/nextstep/edu/step3/racing/InstantEngine.kt:
```diff
@@ -0,0 +1,21 @@
+package camp.nextstep.edu.step3.racing
+
+class InstantEngine(n: Int) : Car.Engine {
```
`InstantEngine` 센스있는 네이밍 입니다. 👏 '👍 
`n` 보다 좀 더 의미있는 이름을 만들어 보면 좋겠습니다.

karian7 reviewed (approved):
모든 피드백을 의도한대로 잘 반영해 주셨습니다. 👏 👏 💯 
깊이 고민하신 내용을 말씀해 주셔서 피드백 드리는데 힘이 되기도 하였습니다.

이번단계는 머지하겠습니다. 다음단계 진행 부탁드리고요, 살짝 진도가 느린데 주말에도 혹시 궁금하신 내용이 있다면 편하게 DM 주세요!
